### DISCUSSION TITLE
Best way to run shell scripts

### DISCUSSION THREAD

**Discussion by rgaiacs** [quarto-dev/quarto-cli#1468] [Q&A] [answered]
When using R Markdown, I can convert the following minimal working example

``````
---
title: "mwe"
author: "Raniere Silva"
date: "2022-07-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{bash}
ls
```
``````

and I will get

![Screenshot from 2022-07-20 21-05-13](https://user-images.githubusercontent.com/1506457/179991155-47df0777-619a-483c-8bb5-fa477254a3c5.png)

With Quarto, the following similar minimal working example

``````
---
title: "mwe"
---

## Quarto

```{bash}
ls
```
``````

raises the following error

```
Starting python3 kernel...Traceback (most recent call last):
  File "/usr/lib/rstudio/bin/quarto/share/jupyter/jupyter.py", line 21, in <module>
    from notebook import notebook_execute, RestartKernel
  File "/usr/lib/rstudio/bin/quarto/share/jupyter/notebook.py", line 16, in <module>
    import nbformat
ModuleNotFoundError: No module named 'nbformat'
Python 3 installation:
  Version: 3.10.5
  Path: /usr/bin/python3
  Jupyter: (None)

Jupyter is not available in this Python installation.
Install with python3 -m pip install jupyter
```

R Markdown was able to handle shell scripts using `system2()` as describe in https://bookdown.org/yihui/rmarkdown-cookbook/eng-bash.html. Do I need to install Jupyter and the Bash kernel? Or there is another solution?

`system2()` works with Quarto! The minimal working example

``````
---
title: "mwe"
---

## Quarto

```{r}
system2("ls", stdout=TRUE)
```
`````` 

produces

![Screenshot from 2022-07-20 21-12-28](https://user-images.githubusercontent.com/1506457/179991078-140262fe-b99d-4a66-8ab2-aa816daf4e0c.png)

**Comment by jjallaire** [comment, marked as answer]
It's using the Jupyter engine by default b/c there are no R code cells. Just add `engine: knitr` to the front-matter and it will work as you expect.

**Comment by rgaiacs** [comment]
@jjallaire Thanks for the reply. Add `engine: knitr` to the front-matter worked. I tried to add it to `_quarto.yml` and it didn't work:

```
project:
  title: "mwe"

editor: visual

knitr: true
```

How do I define the `engine` globally?

**Reply by jjallaire** [reply]
Try this (not 100% sure if it will work though):

```yaml
execute:
   engine: knitr
```

**Reply by rgaiacs** [reply]
Thanks for the answer. It did **not** work.

**Reply by salim-b** [reply]
Tracked in https://github.com/quarto-dev/quarto-cli/issues/3157.

**Comment by stschiff** [comment]
Yes, this also happened to me... I was adding a `bash` block and then it nonsensically switched to the Jupyter engine. 

I then added `engine: knitr` to the frontmatter and it worked, but it's not documented anywhere. For what it's worth, I don't actually find a comprehensive API documentation of the frontmatter at all. There is lots of documentation for specific output documents, but as far as I can see no documentation for general options like this. 

I was also surprised that there doesn't seem to be some kind of command line flag to choose between the two engines. I would have liked something like `quarto render --engine knitr` or so...

Great tool, anyway! But this part is a bit intransparent.

**Reply by jjallaire** [reply]
Docs on engine binding are in the article on execution options: https://quarto.org/docs/computations/execution-options.html#engine-binding

**Reply by stschiff** [reply]
Oh good, I overlooked this. Thanks for the pointer. But note that it doesn't seem exactly right. It looks like it's switching to Jupyter if a python block is discovered, and knitr if an r-block is discovered. But it doesn't say that Jupyter is also chosen with only `bash` blocks. Anyway, thanks!

**Reply by jjallaire** [reply]
Okay, thanks, I've updated the docs to be more clear about this.

**Reply by stschiff** [reply]
Perfect!
